### PR TITLE
feat: Add ModelRegistry component (#775)

### PR DESCRIPTION
**Pull request by dhirajsb**
## Description
This PR adds ModelRegistry as a component to the ODH operator.

## How Has This Been Tested?
Tested with sample DSC CR including the new
```
spec:
  components:
    modelregistry:
      managementState: "Managed"
```
## Merge criteria:
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by dhirajsb** [comment]
@VaishnaviHire any ideas why the e2e tests are failing?

File: tests/e2e/dsc_creation_test.go
Comment (tarilabs) [inline tests/e2e/dsc_creation_test.go:125]: 🤔 
this likely was a typo even in the original/previous writing?

**Review by tarilabs — APPROVED** [review]

File: tests/e2e/dsc_creation_test.go
Comment (dhirajsb) [reply tests/e2e/dsc_creation_test.go:125]: Yeah, this test was a mess. I've cleaned it up.

**Comment by dhirajsb** [comment]
@LaVLaS @VaishnaviHire can you please review this PR. I've verified manually that the odh operator is able to deploy the model-registry-operator and it can create model-registry instances in projects.

**Comment by dhirajsb** [comment]
@VaishnaviHire any ideas why the servicemesh unit test is failing?

**Comment by dhirajsb** [comment]
@zdtsw any pointers to why the unit test is failing for servicemesh test?

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [inline components/modelregistry/modelregistry.go:20]: will you have different path for upstream ODH and downstream RHOAI in `overlays`?

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [inline components/modelregistry/modelregistry.go:57]: need PR ready in midstream with these env variable before sync to RHOAI

File: config/rbac/role.yaml
Comment (zdtsw) [inline config/rbac/role.yaml:100]: no need `delete` or `patch`?

File: config/rbac/role.yaml
Comment (zdtsw) [inline config/rbac/role.yaml:131]: same comments as above

**Comment by zdtsw** [comment]
> @zdtsw any pointers to why the unit test is failing for servicemesh test?

did re-run, looks like a one-off thing

File: components/modelregistry/modelregistry.go
Comment (etirelli) [inline components/modelregistry/modelregistry.go:19]: is there a reason for the component name to be different from the label used in the DSC? "modelregistry" vs "model-registry-operator". While they can be different, it usually causes less confusion if they are the same.

File: components/modelregistry/modelregistry.go
Comment (ykaliuta) [inline components/modelregistry/modelregistry.go:34]: Isn't it incorrect after https://github.com/opendatahub-io/opendatahub-operator/pull/659 ?

File: components/modelregistry/modelregistry.go
Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:20]: Is that needed? I thought that ODH/RHOAI differences were handled through different image names and image env variables.

File: components/modelregistry/modelregistry.go
Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:57]: Thanks, can you point me to a similar PR for that? I'll open one for model-registry.

File: config/rbac/role.yaml
Comment (dhirajsb) [reply config/rbac/role.yaml:100]: So, I saw these errors for missing permissions for daemonsets and deploymentconfigs and added them based on the errors. These resource dependencies are not coming from model-registry though. So, it looks like they were missing in the original incubation branch.

File: components/modelregistry/modelregistry.go
Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:19]: Good question @etirelli . I was going by what DSP has done. It looks like there are multiple places where components are referenced, and this one happens to refer to the github project, which is the `model-registry-operator` project. The config property in the DSC CR is called modelregistry from `apis/datasciencecluster/v1/datasciencecluster_types.go`.

File: components/modelregistry/modelregistry.go
Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:34]: I guess the issue is that DevFlags could be nil, is that correct? There are still several components where the len(x.DevFlags.Manifests) check is being used. I suppose the replacement should be:
```
if m.DevFlags != nil && len(m.DevFlags.Manifests) != 0 {
```

Comment (ykaliuta) [reply components/modelregistry/modelregistry.go:34]: For those components probably the check is implemented before the function call https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/components/codeflare/codeflare.go#L71

Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:34]: Good catch. I must have started the PR before this change went in. Added check in Reconcile like other components.

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [reply components/modelregistry/modelregistry.go:20]: images for ODH and RHOAI are different, manifests can be exactly the same or totally different.
take dsp and  dashboard as 2 examples.
but if modelreg is decided to have this path, i am all fine with the code.

File: config/rbac/role.yaml
Comment (zdtsw) [reply config/rbac/role.yaml:100]: ok, I see. I thought modelreg requires such resources back.
I did the cleanup for `daemonsets` and `deploymentconfigs` a while a go, after confirmed with workbench team these are not in use any more.
do you still have the error logs where they showed up? I would like to take a look if this is a bug from my original PR or not.

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [reply components/modelregistry/modelregistry.go:57]: i think the standard workflow is a jira to devops team to deal with any new component in RHOAI.
there was some docs about the process but i dont recall any detail.

**Comment by zdtsw** [comment]
@dhirajsb do you have any related jira issue can link in the description?

File: config/rbac/role.yaml
Comment (dhirajsb) [reply config/rbac/role.yaml:100]: The errors I got on Dec 19 were:
```
W1220 05:01:48.998743 1 reflector.go:424] go/pkg/mod/k8s.io/client-go@v0.26.0/tools/cache/reflector.go:169: failed to list *v1.DeploymentConfig: deploymentconfigs.apps.openshift.io is forbidden: User "system:serviceaccount:opendatahub-operator-system:opendatahub-operator-controller-manager" cannot list resource "deploymentconfigs" in API group "apps.openshift.io" at the cluster scope
E1220 05:01:48.998767 1 reflector.go:140] go/pkg/mod/k8s.io/client-go@v0.26.0/tools/cache/reflector.go:169: Failed to watch *v1.DeploymentConfig: failed to list *v1.DeploymentConfig: deploymentconfigs.apps.openshift.io is forbidden: User "system:serviceaccount:opendatahub-operator-system:opendatahub-operator-controller-manager" cannot list resource "deploymentconfigs" in API group "apps.openshift.io" at the cluster scope
W1220 05:01:49.005990 1 reflector.go:424] go/pkg/mod/k8s.io/client-go@v0.26.0/tools/cache/reflector.go:169: failed to list *v1.DaemonSet: daemonsets.apps is forbidden: User "system:serviceaccount:opendatahub-operator-system:opendatahub-operator-controller-manager" cannot list resource "daemonsets" in API group "apps" at the cluster scope
E1220 05:01:49.006215 1 reflector.go:140] go/pkg/mod/k8s.io/client-go@v0.26.0/tools/cache/reflector.go:169: Failed to watch *v1.DaemonSet: failed to list *v1.DaemonSet: daemonsets.apps is forbidden: User "system:serviceaccount:opendatahub-operator-system:opendatahub-operator-controller-manager" cannot list resource "daemonsets" in API group "apps" at the cluster scope
```

Maybe you fixed that after I added it to this PR.

**Comment by dhirajsb** [comment]
> @dhirajsb do you have any related jira issue can link in the description?

@zdtsw The jira epic is RHOAIENG-442. We can add downstream issues to that epic as needed. Our priority right now is to get the component into incubation upstream and work on downstream productization issues next.

File: components/modelregistry/modelregistry.go
Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:57]: Thanks, I'll ping devops team about it.

**Comment by dhirajsb** [comment]
@zdtsw looks like the e2e job is failing because etcd startup failed. Is that test env maintained by devops?

**Comment by dhirajsb** [comment]
It looks like the following tests are failing in e2e tests:
```
                --- FAIL: TestOdhOperator/create_Opendatahub_components/e2e-test-dsc/Validate_all_deployed_components/Validate_CodeFlare (120.01s)
                --- FAIL: TestOdhOperator/create_Opendatahub_components/e2e-test-dsc/Validate_all_deployed_components/Validate_ModelMeshServing (120.05s)
```
ModelRegistry test is working.

**Comment by ykaliuta** [comment]
> It looks like the following tests are failing in e2e tests:
> 
> ```
>                 --- FAIL: TestOdhOperator/create_Opendatahub_components/e2e-test-dsc/Validate_all_deployed_components/Validate_CodeFlare (120.01s)
>                 --- FAIL: TestOdhOperator/create_Opendatahub_components/e2e-test-dsc/Validate_all_deployed_components/Validate_ModelMeshServing (120.05s)
> ```
> 
> ModelRegistry test is working.

Your PR changes test's logic https://github.com/opendatahub-io/opendatahub-operator/pull/776/files#diff-ce0ca88eff51d2d34313ca907cf61e0f9c6b4bdc3756757628e1a50699f01cc8L243-L254  (see require.NoError() for both err == nil and err != nil)

The change is unrelated to "Add ModelRegistry component". With the current workflow probably it's better to have a separate issue/PR to fix the test but leave the code unchanged in the PR. Same about some other changes there.

**Comment by dhirajsb** [comment]
> Your PR changes test's logic https://github.com/opendatahub-io/opendatahub-operator/pull/776/files#diff-ce0ca88eff51d2d34313ca907cf61e0f9c6b4bdc3756757628e1a50699f01cc8L243-L254 (see require.NoError() for both err == nil and err != nil)

Right, that change seems to be working for all the other components. Since it's just code cleanup for the first `if err != nil` check. The second check should be the same way, unless for CodeFlare and ModelMesh components it was returning a non-nil err when component is removed. 
I'll test by reverting that change for those two components and verify. Thanks.

File: tests/e2e/dsc_creation_test.go
Comment (ykaliuta) [inline tests/e2e/dsc_creation_test.go:164]: I repeat my question, why do you push these changes in this PR? It's unrelated to adding the component.

File: tests/e2e/dsc_deletion_test.go
Comment (ykaliuta) [inline tests/e2e/dsc_deletion_test.go:81]: Here as well, it has nothing to do with adding the component.

File: tests/e2e/dsc_creation_test.go
Comment (dhirajsb) [reply tests/e2e/dsc_creation_test.go:164]: Adding the model registry tests pushed the cyclomatic complexity of the code beyond the linter limit. So I had to clean up un-necessary code like`if nil` check before calling `require.NoError`. It won't pass the linter without simplifying the obviously bad code. 
The tests for all components for `Disabled` state used some useless code that does something that will never be true (`if err == nil` check before calling `require.NoError`). It looked like it should be checking for an error instead, which seems to pass for other components, except CodeFlare and ModelMeshServing. So I modified the test code to do what it was doing before, i.e. use `err == nil` with `require.NoError`.

File: tests/e2e/dsc_deletion_test.go
Comment (dhirajsb) [reply tests/e2e/dsc_deletion_test.go:81]: Same answer, go linter version changed and it was throwing errors here. So, I had to clean up the code syntax.

File: tests/e2e/dsc_creation_test.go
Comment (dhirajsb) [reply tests/e2e/dsc_creation_test.go:164]: I went back to the git history for this code and at one point the test code flipped from looking for errors for disabled components, to not doing anything at all. I'm changing it back to what it should be doing for disabled components, i.e. using `require.Error`.

Comment (ykaliuta) [reply tests/e2e/dsc_creation_test.go:164]: I added your tests to the incubation branch and `make lint` runs just fine. Probably you can also disable it with annotations if it causes problems with integration.

I did not claim that the tests are correct, but fixing them sounds like a separate task, not part of adding a component.

Comment (ykaliuta) [reply tests/e2e/dsc_creation_test.go:164]: Ok, taking my words back :) Updated golangci-lint to 1.55.2 and it fails, but actually in many more places.

Comment (ykaliuta) [reply tests/e2e/dsc_creation_test.go:164]: Here is your tests in github https://github.com/opendatahub-io/opendatahub-operator/pull/815

**Comment by dhirajsb** [comment]
Here are the linter errors from past ci builds for this PR to show why the supposedly unrelated code changes were required:

* cyclomatic complexity: https://github.com/opendatahub-io/opendatahub-operator/actions/runs/7096844413
* if-return: redundant if: https://github.com/opendatahub-io/opendatahub-operator/actions/runs/7532005039

Looks like golanci-lint check changed to not throw those errors anymore. IMHO it's a bad engineering practice to disable code checks instead of fixing the code. As has been requested, I've created another PR #817 for the code cleanup and error handling fixes. 

I'll rebase this PR after that PR has been merged.

**Comment by ykaliuta** [comment]
> Here are the linter errors from past ci builds for this PR to show why the supposedly unrelated code changes were required:
> 
> * cyclomatic complexity: https://github.com/opendatahub-io/opendatahub-operator/actions/runs/7096844413
> * if-return: redundant if: https://github.com/opendatahub-io/opendatahub-operator/actions/runs/7532005039

I see, thank you a lot for clarification!

> Looks like golanci-lint check changed to not throw those errors anymore.

Yes, the code I propose https://github.com/opendatahub-io/opendatahub-operator/pull/815 with just adding your tests passes both linter and e2e.

> IMHO it's a bad engineering practice to disable code checks instead of fixing the code.

I cannot agree more! And it's not needed anymore, as you mentioned, so we can forget about that :)

> As has been requested, I've created another PR #817 for the code cleanup and error handling fixes.
> 
> I'll rebase this PR after that PR has been merged.

Thank you a lot! It's an ideal way, probably. The only thing I do not like is that this creates extra dependency for your PR which can slow down its integration. Why don't you leave fixing tests for the operator and focus only on the component integration?

Anyway, if  #817 can be merged in time fine for me.

**Comment by dhirajsb** [comment]
The [lint error](https://github.com/opendatahub-io/opendatahub-operator/actions/runs/7658544911/job/20871720673?pr=776) in dsc_deleteion_test.go is back:
```
Warning: if-return: redundant if ...; err != nil check, just return error instead. (revive)
```
That needs to be changed to what I had done before.

**Comment by dhirajsb** [comment]
We are working on creating a release for model-registry, and will replace all the references to main branch with this release. So please hold off merging this PR until that's tomorrow.

**Comment by VaishnaviHire** [comment]
> We are working on creating a release for model-registry, and will replace all the references to main branch with this release. So please hold off merging this PR until that's tomorrow.

For `incubation` branch it is fine to use main branch references. We update the release versions in `odh` release branch, E.g  dashboard release for `odh-2.4` branch https://github.com/opendatahub-io/opendatahub-operator/blob/odh-2.4.0/get_all_manifests.sh#L11

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [reply components/modelregistry/modelregistry.go:57]: could not find this part has been done in CPaaS, otherwise, i am fine with the change in PR.

**Review by zdtsw — APPROVED** [review]
/lgtm

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [reply components/modelregistry/modelregistry.go:57]: to not let CPaaS become a blocker for merging to incubation, we can leave this part for later commits

Comment (dhirajsb) [reply components/modelregistry/modelregistry.go:57]: We are targeting the 2.8 release. We'll add these variables to that release in CPaaS.

**Comment by dhirajsb** [comment]
@VaishnaviHire, we'll stay on the model registry `main` version on `incubation` branch. Please merge this PR. TIA

**Pull request merged**
